rabbitmq/consumer.go

```diff
@@ -26,51 +26,85 @@ import (
 	"github.com/streadway/amqp"
 )
 
-// A consumer that is works with Handler interface
-// It needs a RabbitMQClient to work with and is started with the Run() method
-type RabbitMQConsumer struct {
+type ConsumerConfig struct {
+	// PrefetchCount configures how many in-flight "deliveries" are available to the consumer to ack/nack.
+	// ref: https://www.rabbitmq.com/consumer-prefetch.html
+	// There's no default value for the reason that it's very easy to misuse RMQ and have multiple consumers
+	// with too many deliveries in flight which results into badly distributed work load and high memory footprint
+	// of the consumers.
+	PrefetchCount int
+}
+
+type Consumer struct {
 	client  *RabbitMQClient
-	done    chan bool
 	handler Handler
 	logger  logger.StructuredLogger
 	metric  Metric
+	cfg     ConsumerConfig
 }
 
 func NewConsumer(
 	client *RabbitMQClient,
 	handler Handler,
 	logger logger.StructuredLogger,
 	metric Metric,
-) *RabbitMQConsumer {
-	return &RabbitMQConsumer{
+	cfg ConsumerConfig,
+) *Consumer {
+	return &Consumer{
 		client:  client,
-		done:    make(chan bool),
 		handler: handler,
 		logger:  logger,
 		metric:  metric,
+		cfg:     cfg,
 	}
 }
 
-func (c *RabbitMQConsumer) Run(ctx context.Context) error {
+func (c *Consumer) Run(ctx context.Context) error {
+	channel, err := c.client.CreateChannel(ctx)
+	if err != nil {
+		return stacktrace.Propagate(err, "failed to create a RMQ channel")
+	}
+
+	ctx, cancelFunc := context.WithCancel(ctx)
+	defer cancelFunc()
+
+	closeCh := channel.NotifyClose(make(chan *amqp.Error))
+
 	go func() {
-		<-ctx.Done()
-		c.logger.Info("Received context cancel. Going to close rabbit connections.")
-		_ = c.client.channel.Cancel(c.handler.GetConsumerTag(), false)
+		select {
+		case rmqErr := <-closeCh:
+			cancelFunc()
+			c.logger.Warn(
+				"RMQ closed the connection",
+				zap.String("reason", rmqErr.Reason),
+				zap.Int("code", rmqErr.Code),
+				zap.Bool("recover", rmqErr.Recover),
+				zap.Bool("server", rmqErr.Server),
+			)
+			return
+		case <-ctx.Done():
+			c.logger.Info("Received context cancel. Going to close RMQ connections.")
+			_ = channel.Cancel(c.handler.GetConsumerTag(), false)
+
+			if !c.handler.WaitToConsumeInflight() {
+				_ = channel.Close()
+			}
 
-		if !c.handler.WaitToConsumeInflight() {
-			c.client.channel.Close()
+			c.logger.Info("RMQ consumer stopped.")
+			_ = c.client.Close()
```

Comment:
Would it be better if we log these potential errors, as that's still one way of handling them?

Reply:
Do we have if the client has been closed "incorrectly" if we're already stopping?

Reply:
Probably not usually but maybe it's still worth it if we fail to cancel the channel, i.e. getting weird behaviour (eventually) and disregarding the errors might result in harder to find issues?